### PR TITLE
Consolidate CSP nonce sanitization into shared module

### LINKED ISSUE #2582: Follow-up: CSP nonce sanitization consolidation and validation policy
## Context
PR #2418 contains two non-blocking review topics that we are deferring so the PR can merge while @abanoub is away.

## Follow-up items
1. Consolidate duplicate nonce sanitization logic
- Current duplicate implementations:
  - `packages/react-on-rails/src/RenderUtils.ts`
  - `packages/react-on-rails-pro/src/utils.ts`
- Decide whether to extract a shared helper and where it should live given package boundaries.
- Ensure behavior stays aligned across OSS and Pro paths.

2. Decide malformed nonce handling policy
- Decide between strict reject/logging vs current sanitize-then-validate behavior in Pro `sanitizeNonce`.
- Document the chosen policy and add/adjust tests as needed.

## Related PR discussions
- https://github.com/shakacode/react_on_rails/pull/2418#discussion_r2903735101
- https://github.com/shakacode/react_on_rails/pull/2418#discussion_r2903735102

### PR DESCRIPTION
**Pull request by justin808**
## Summary
- Extracted duplicate `sanitizeNonce` logic from OSS `RenderUtils.ts` and Pro `utils.ts` into a single `react-on-rails/sanitizeNonce` module
- Pro package now imports from OSS instead of maintaining its own copy
- Added comprehensive tests (11 cases) covering valid nonces, injection attempts, edge cases
- Documented the sanitize-then-validate policy: malformed nonces are silently dropped (no nonce attribute emitted) rather than rejecting the render or logging sensitive values

Fixes #2582

## Test plan
- [x] OSS `sanitizeNonce.test.ts` — 11 tests pass
- [x] OSS `buildConsoleReplay.test.ts` — still passes (uses `wrapInScriptTags` which now imports the shared module)
- [x] Pro `injectRSCPayload.test.ts` — nonce sanitization and valid nonce tests pass
- [x] Pro `transformRSCStreamAndReplayConsoleLogs.test.ts` — nonce tests pass
- [x] Full `pnpm run build` succeeds
- [x] `pnpm run type-check` passes
- [x] Prettier formatting verified

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Touches CSP nonce handling used to generate/attach `<script>` nonces; mistakes could weaken CSP protections or break script execution. Scope is small and behavior is now unit-tested, reducing regression risk.
> 
> **Overview**
> Centralizes CSP nonce sanitization by introducing `packages/react-on-rails/src/sanitizeNonce.ts`, wiring `wrapInScriptTags` to use it, and exporting it as `react-on-rails/@internal/sanitizeNonce` for reuse.
> 
> Removes the duplicated `sanitizeNonce` implementation from `react-on-rails-pro/src/utils.ts` and updates Pro RSC helpers (`injectRSCPayload`, `transformRSCStreamAndReplayConsoleLogs`) to import the shared sanitizer. Adds a new Jest suite covering nonce edge cases/injection attempts, and includes small maintenance updates to `.lychee.toml` exclusions and a formatting fix in `PROJECTS.md`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 37d18b4860f89dd0f8f802a88db3d643c812d384. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved CSP nonce handling by centralizing sanitization and preserving previous validation/behavior.

* **Chores**
  * Consolidated nonce validation into a shared export and adjusted streaming helpers for compatibility with newer async runtimes.

* **Tests**
  * Added unit tests covering nonce sanitization, validation, and injection-edge cases.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -34,6 +34,10 @@ After a release, run `/update-changelog` in Claude Code to analyze commits, writ
   [PR 2816](https://github.com/shakacode/react_on_rails/pull/2816) by
   [justin808](https://github.com/justin808).
 
+#### Changed
+
+- **Consolidated CSP nonce sanitization**: Extracted the duplicate `sanitizeNonce` logic from `RenderUtils.ts` (OSS) and `utils.ts` (Pro) into a single shared `sanitizeNonce` module in the OSS package (`react-on-rails/sanitizeNonce`). The Pro package now imports from the OSS package instead of maintaining its own copy. Behavior is unchanged: malformed nonces are silently dropped (sanitize-then-validate policy). [PR 2828](https://github.com/shakacode/react_on_rails/pull/2828) by [justin808](https://github.com/justin808). Fixes [Issue 2582](https://github.com/shakacode/react_on_rails/issues/2582).
```
[the remark this answers is not in the document]
Comment (justin808) [reply CHANGELOG.md]: Addressed. The changelog now references [PR 2828](https://github.com/shakacode/react_on_rails/pull/2828) directly (no placeholder link remains).

File: packages/react-on-rails/src/sanitizeNonce.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
+/**
+ * Sanitizes a CSP nonce to prevent attribute injection attacks.
+ *
+ * Policy: sanitize-then-validate. Characters outside the base64/base64url alphabet
+ * are stripped first, then the result is validated against the expected nonce pattern.
+ * If the sanitized value does not match, `undefined` is returned and no nonce
+ * attribute will be emitted — the render proceeds without a nonce rather than
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails/src/sanitizeNonce.ts:7]: Addressed in a73a2e6d. I clarified `sanitizeNonce` docs to state that sanitize-then-validate may return a stripped value when it still matches the nonce pattern.

File: packages/react-on-rails/tests/sanitizeNonce.test.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    expect(sanitizeNonce('abc+def/ghi')).toBe('abc+def/ghi');
+  });
+
+  it('strips and rejects a nonce containing attribute injection', () => {
+    expect(sanitizeNonce('abc123" onload=alert(1)')).toBeUndefined();
+  });
+
+  it('strips dangerous characters from injection attempts', () => {
+    // After stripping non-base64 chars, the remaining alphanumeric string is valid
+    expect(sanitizeNonce('"><script>alert(1)</script>')).toBe('scriptalert1/script');
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails/tests/sanitizeNonce.test.ts]: Addressed in a73a2e6d. Updated the test comment to explicitly call out that the resulting value is attribute-safe but may be nonce-like/meaningless in practice unless it matches a server-generated nonce.

[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails/tests/sanitizeNonce.test.ts]: Addressed in a73a2e6d by renaming this case to `strips invalid chars from injection input and returns the remaining nonce text`, plus a clearer safety/semantics comment.

File: packages/react-on-rails/src/sanitizeNonce.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+ * failing or logging potentially sensitive values.
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails/src/sanitizeNonce.ts]: Addressed in a73a2e6d. The inline docs now include the explicit edge case that stripped output can still pass pattern validation and be returned.

File: CHANGELOG.md [hunk 5]
```diff
@@ -61,6 +61,10 @@ After a release, run `/update-changelog` in Claude Code to analyze commits, writ
 - **[Pro]** **Minimum `async` gem version bumped to 2.29**: The streaming helper now requires `async >= 2.29` (previously `>= 2.6`) due to the migration from `Async::Variable` to `Async::Promise`. If your Gemfile pins the `async` gem below 2.29, you will need to update it before upgrading React on Rails Pro. Run `bundle update async` to pick up the new minimum.
   [PR 2832](https://github.com/shakacode/react_on_rails/pull/2832) by [justin808](https://github.com/justin808).
 
+#### Changed
+
+- **[Pro]** **Migrated from `Async::Variable` to `Async::Promise`**: The streaming helper internals now use `Async::Promise` for async v2.29+ compatibility while preserving pre-first-chunk error propagation behavior. [PR 2832](https://github.com/shakacode/react_on_rails/pull/2832) by [justin808](https://github.com/justin808). Fixes [Issue 2563](https://github.com/shakacode/react_on_rails/issues/2563).
```
[the remark this answers is not in the document]
Comment (justin808) [reply CHANGELOG.md]: Addressed: I removed the duplicate `#### Changed` block and the duplicated PR 2832 entry from `CHANGELOG.md`. The PR 2828 nonce consolidation change is intentionally not added to the changelog because it is an internal refactor rather than a user-visible change.

[the remark this answers is not in the document]
Comment (justin808) [reply CHANGELOG.md]: Addressed: I removed the duplicate `#### Changed` block and the duplicated PR 2832 entry from `CHANGELOG.md`. The PR 2828 nonce consolidation change is intentionally not added to the changelog because it is an internal refactor rather than a user-visible change.

File: packages/react-on-rails/package.json [hunk 6]
```diff
@@ -48,6 +48,7 @@
     "./serverRenderUtils": "./lib/serverRenderUtils.js",
     "./buildConsoleReplay": "./lib/buildConsoleReplay.js",
     "./ReactDOMServer": "./lib/ReactDOMServer.cjs",
+    "./sanitizeNonce": "./lib/sanitizeNonce.js",
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails/package.json]: Addressed. The export is now internal-only (`./@internal/sanitizeNonce`) and Pro imports use `react-on-rails/@internal/sanitizeNonce`, so this does not add a new top-level public API.

File: packages/react-on-rails/src/sanitizeNonce.ts [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+ * If the sanitized value does not match, `undefined` is returned and no nonce
+ * attribute will be emitted — the render proceeds without a nonce rather than
+ * failing or logging potentially sensitive values. Note that if stripping yields
+ * a string that still matches the base64/base64url pattern, that stripped value
+ * is returned.
+ *
+ * CSP nonces should be base64 or base64url strings with optional trailing `=` padding.
+ */
+export default function sanitizeNonce(nonce?: string): string | undefined {
+  const nonceWithAllowedCharsOnly = nonce?.replace(/[^a-zA-Z0-9+/=_-]/g, '');
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails/src/sanitizeNonce.ts:15]: Addressed. There is explicit coverage for this asymmetry: `sanitizeNonce('abc=def')` now expects `undefined` in `sanitizeNonce.test.ts`.

File: packages/react-on-rails/tests/sanitizeNonce.test.ts [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+  it('passes through a nonce with plus and slash (standard base64)', () => {
+    expect(sanitizeNonce('abc+def/ghi')).toBe('abc+def/ghi');
+  });
+
+  it('strips and rejects a nonce containing attribute injection', () => {
+    expect(sanitizeNonce('abc123" onload=alert(1)')).toBeUndefined();
+  });
+
+  it('strips invalid chars from injection input and returns the remaining nonce text', () => {
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails/tests/sanitizeNonce.test.ts:36]: Addressed. The test comment now explicitly states the emitted nonce-like value will not match the server CSP nonce, so the script remains blocked by CSP.

File: CHANGELOG.md [hunk 9]
```diff
@@ -61,6 +61,11 @@ After a release, run `/update-changelog` in Claude Code to analyze commits, writ
...
+
+- **Consolidated CSP nonce sanitization**: Extracted the duplicate `sanitizeNonce` logic from `RenderUtils.ts` (OSS) and `utils.ts` (Pro) into a single shared `sanitizeNonce` module in the OSS package (`react-on-rails/sanitizeNonce`). The Pro package now imports from the OSS package instead of maintaining its own copy. Behavior is unchanged: sanitize-then-validate is applied, and values that still fail nonce-pattern validation after sanitization are omitted. [PR 2828](https://github.com/shakacode/react_on_rails/pull/2828) by [justin808](https://github.com/justin808). Fixes [Issue 2582](https://github.com/shakacode/react_on_rails/issues/2582).
```
[the remark this answers is not in the document]
Comment (justin808) [reply CHANGELOG.md]: Addressed. The duplicate `#### Changed` block and duplicated PR 2832 entry were removed from `CHANGELOG.md`; only the original section remains.

File: packages/react-on-rails/src/sanitizeNonce.ts [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+  return nonceWithAllowedCharsOnly?.match(/^[a-zA-Z0-9+/_-]+={0,2}$/)?.[0];
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails/src/sanitizeNonce.ts:16]: Leaving as-is. In JavaScript, a trailing `-` in a character class is literal; we kept this form to preserve exact behavior parity with prior implementations.

[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails/src/sanitizeNonce.ts:16]: Intentional no-change for this PR. This change set is consolidation with behavior parity, so we preserved sanitize-then-validate semantics and documented them in source/tests.

File: packages/react-on-rails/src/sanitizeNonce.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails/src/sanitizeNonce.ts:15]: Addressed. Added explicit test coverage for middle `=` (`sanitizeNonce('abc=def') -> undefined`) in `sanitizeNonce.test.ts`.

File: packages/react-on-rails/src/sanitizeNonce.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails/src/sanitizeNonce.ts:16]: No change needed. Accepting the base64/base64url superset is intentional and preserves existing behavior; strict alphabet separation is out of scope for this refactor.

File: packages/react-on-rails/tests/sanitizeNonce.test.ts [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    // The output is attribute-safe and cannot inject markup; it may still be a nonce-like
+    // string that does not match any server-generated CSP nonce in practice.
+    expect(sanitizeNonce('"><script>alert(1)</script>')).toBe('scriptalert1/script');
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails/tests/sanitizeNonce.test.ts:40]: Addressed. The inline comment now notes that even when a nonce-like value is emitted, it still does not match the server CSP nonce, so the script stays blocked.

File: packages/react-on-rails/src/sanitizeNonce.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails/src/sanitizeNonce.ts:16]: No code change. The strip-and-accept path is already covered (`'abc 123' -> 'abc123'`), and mixed alphabet acceptance is intentional behavior parity here.

File: packages/react-on-rails/tests/sanitizeNonce.test.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails/tests/sanitizeNonce.test.ts:40]: Design suggestion acknowledged. Reject-on-any-strip is a broader policy change; this PR intentionally preserves the existing sanitize-then-validate behavior.

File: packages/react-on-rails/src/sanitizeNonce.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails/src/sanitizeNonce.ts:16]: Intentional no-change for this PR. Validate-then-reject is a viable alternative policy, but this refactor explicitly preserves previous behavior and documents current semantics.

File: packages/react-on-rails/tests/sanitizeNonce.test.ts [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+  it('strips and rejects a nonce containing attribute injection', () => {
+    expect(sanitizeNonce('abc123" onload=alert(1)')).toBeUndefined();
+  });
+
+  it('strips invalid chars from injection input and returns the remaining nonce text', () => {
+    // The output is attribute-safe and cannot inject markup; it may still be a nonce-like
+    // string that does not match any server-generated CSP nonce in practice, so the script
+    // tag would still be blocked by CSP.
+    expect(sanitizeNonce('"><script>alert(1)</script>')).toBe('scriptalert1/script');
+  });
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails/tests/sanitizeNonce.test.ts:41]: Kept current wording. The test body/comment already state the output and CSP effect; treating this as naming preference rather than a blocking issue.

File: packages/react-on-rails/package.json [hunk 13]
```diff
@@ -49,6 +49,7 @@
     "./buildConsoleReplay": "./lib/buildConsoleReplay.js",
     "./ReactDOMServer": "./lib/ReactDOMServer.cjs",
     "./serverRenderReactComponent": "./lib/serverRenderReactComponent.js",
+    "./@internal/sanitizeNonce": "./lib/sanitizeNonce.js",
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails/package.json:52]: Intentional. `@internal` communicates an unstable internal surface for package consumers, and this internal refactor is not changelog-worthy under repo policy.

**Comment by justin808** [comment]
Address-review checkpoint posted at `2026-04-09T08:02:52Z`.

Scan window: full PR history (`2026-03-24T06:50:15Z` → `2026-04-08T07:37:26Z`) because no previous `<!-- address-review-summary -->` comment existed.
Cutoff mode: default workflow cutoff (no `check all reviews` override used).

### Mattered
- Confirmed and closed addressed blocking items from earlier review passes:
  - `./sanitizeNonce` export concern was handled by using `./@internal/sanitizeNonce`.
  - `abc=def` middle-padding behavior now has explicit test coverage.
  - CSP-mismatch implication is now called out in sanitizer test comments.
  - Duplicate `CHANGELOG.md` section/commentary was removed.
- Policy-change suggestions (sanitize-then-validate vs validate-then-reject) were reviewed and intentionally deferred: this PR remains behavior-preserving consolidation.

### Skipped
- Regex character-class formatting nit (`-` placement) with no behavioral impact.
- Test-name wording preferences where test body/comments already capture behavior.
- General notes about `@internal` discoverability/changelog annotation for internal-only surfaces.
- Duplicate/status-only bot summary comments that did not add new actionable deltas.

Follow-up issue: none created (no deferred must-fix items).

Future full-PR scans should start after this comment unless `check all reviews` is explicitly requested.
